aes: add initial AES module

xobs opened the pull request:
This adds initial support for an AES cipher module.  This
implementation supports only a subset of AES modes, namely
ECB, CBC, and CTR modes.  These are part of the TinyAES library.

Example usage adapted from the PyCrypto README for AES:

```
>>> import aesio
>>>
>>> key = b'Sixteen byte key'
>>> cipher = aesio.AES(key, aesio.MODE_ECB)
>>> output = bytearray(16)
>>> cipher.encrypt_into(b'Circuit Python!!', output)
>>> output
bytearray(b'E\x14\x85\x18\x9a\x9c\r\x95>\xa7kV\xa2`\x8b\n')
>>>
```

 This key is 16-bytes, so it uses AES128.  If your key is 24- or 32-
 bytes long, it will switch to AES192 or AES256 respectively.

This adds additional functions to keep the number of reallocations to a minimum.  For example, `rekey()` can be used to reset a CBC stream or set a new IV.  There are also functions to perform operations on existing buffers:

```
>>> out_buffer = bytearray(16)
>>> in_buffer = cipher.encrypt(b'Circuit Python!!')
>>> in_buffer
bytearray(b'E\x14\x85\x18\x9a\x9c\r\x95>\xa7kV\xa2`\x8b\n')
>>> cipher.decrypt_into(in_buffer, out_buffer)
>>> out_buffer
bytearray(b'Circuit Python!!')
>>>
```

 This has been tested with many of the official NIST test vectors,
 such as those used in `pycryptodome` at
 https://github.com/Legrandin/pycryptodome/tree/39626a5b01ce5c1cf51d022be166ad0aea722177/lib/Crypto/SelfTest/Cipher/test_vectors/AES

 CTR has not been tested as NIST does not provide test vectors for it.

xobs commented:
Note that this uses the capitalised module name `Crypto` to be compatible with the pycrypto module at https://www.dlitz.net/software/pycrypto/api/current/Crypto.Cipher.AES-module.html

I've added documentation, but I'm not sure if I documented it correctly.

ladyada commented:
@brentru fyi, maybe there's other crypto algs we could add that you needed? https://www.dlitz.net/software/pycrypto/api/current/Crypto-module.html

Dar-Scott commented:
The protocol used by the [Covid Watch](covid-watch.org) anonymous exposure notification app to fight Covid-19 uses **SHA256** and **Ed25519 sig/ver**. If a "kid-mode" of the app is implemented in CLUE or CircuitPlayground Bluefruit, those would be handy. I don't know what would be needed to implement the Apple/Google protocol; that might be a little different.

tannewt commented:
> Note that this uses the capitalised module name `Crypto` to be compatible with the pycrypto module at https://www.dlitz.net/software/pycrypto/api/current/Crypto.Cipher.AES-module.html
> 
> I've added documentation, but I'm not sure if I documented it correctly.

Let's not match pycrypto at the package level. Lets only have the aes module API match. That way we can turn on and off crypto algorithms at the module level. So, the cross-platform import would be:

```python
import sys
if sys.implementation.name == "circuitpython":
  import aes as AES
else:
  from Crypto.Cipher import AES
```

It'd be awesome to have port-specific implementations that use hardware peripherals for the computation too. This is a great start so everything can have it.

tannewt commented on shared-bindings/crypto/aes.c:
```diff
@@ -0,0 +1,254 @@
+//| :class:`AESCipher` -- Encrypt and decrypt AES streams
+//| =====================================================
+//|
+//| An object that decodes MP3 files for playback on an audio device.
```
Update this.

tannewt commented on shared-bindings/crypto/__init__.c:
```diff
@@ -0,0 +1,86 @@
+//| ==================================================================
+//|
+//| .. module:: Crypto.Cipher
+//|   :synopsis: Embedded implementation of PyCrypto
```
We don't want to mimic pycrypto. I got the wrong python module. (PyCrypto was last released in 2013.) It looks like `cryptography` is the recommended package: https://docs.python-guide.org/scenarios/crypto/

Instead, I think we want our own APIs that only feature `encrypt_into` and `decrypt_into` that both take two arguments, the source and destination buffers. They can be the same or different. Omitting the methods that return new memory means folks are less likely to generate a bunch of garbage they don't want.

xobs replied:
Is there an API reference for that library?  I used to have a simpler api with an `encrypt()` and `decrypt()` call, but I threw that away to create API compatibility with PyCrypto.

What should the API look like when it comes to instantiation, key and IV management, and actual usage?

tannewt replied:
Yup! It is here: https://cryptography.io/en/latest/

It's not super clear to me how you arbitrarily assemble pieces from the `hazmat` layer.

hierophect commented:
@tannewt I'm interested in this idea of port-level implementations. A lot of the STM32s have optional Crypto processors as a sort of mini-upgrade to the chip - the F415 is one such upgrade to the F405, with AES support.

xobs commented:
@hierophect some parts in the nRF52 family, for example, also contain crypto accelerators for certain primitives. For example, AES128.

@tannewt that new library looks like it would be complicated to implement.  They decouple `modes` from `algorithms` from `encryptors` from `backends`.  The best example appears to be from https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/#cryptography.hazmat.primitives.ciphers.Cipher:

```python
>>> import os
>>> from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
>>> from cryptography.hazmat.backends import default_backend
>>> backend = default_backend()
>>> key = os.urandom(32)
>>> iv = os.urandom(16)
>>> cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=backend)
>>> encryptor = cipher.encryptor()
>>> ct = encryptor.update(b"a secret message") + encryptor.finalize()
>>> decryptor = cipher.decryptor()
>>> decryptor.update(ct) + decryptor.finalize()
b'a secret message'
```

hathach commented:
I chime in a bit, since I added a Arduino library supported for ARM CryptoCell CC310 on nrf52840. Unlike AES on all nrf52 which is shard with SoftDevice, CC310 subsystem is separated and available to use, it is more powerful and could do lost of crypto with arm/nordic a.lib . I have used it for computing the ECC DH 

https://github.com/adafruit/Adafruit_nRFCrypto 

https://infocenter.nordicsemi.com/topic/ps_nrf52840/cryptocell.html?cp=4_0_0_5_5

![image](https://user-images.githubusercontent.com/249515/80679055-5916d080-8ae6-11ea-9d3c-abe224ffce9d.png)

hathach commented:
PS: I think mbedtls/wolfssl does adding CC310 support on nrf52840 as back-end for their API as well, we can just use them if needed. Sorry if I am out of context here for this PR :D 

https://github.com/ARMmbed/mbed-os/tree/master/features/cryptocell/FEATURE_CRYPTOCELL310

https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/src/port/arm/cryptoCell.chttps://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/src/port/arm/cryptoCell.c

tannewt commented:
> @tannewt that new library looks like it would be complicated to implement.

Agree. Go ahead and do an API you think is best. (Remember that it shouldn't return new bytestrings because that forces an allocation.)

xobs commented:
I switched to using our own API under the `aes` namespace.  As requested, I added functions such as `decrypt_into()` and `encrypt_into()` that will operate on a predefined buffer.  I also added `rekey()` which can be used to set a new key or a new IV without reallocating the object.

tannewt commented on shared-bindings/aes/aes.c:
```diff
@@ -0,0 +1,355 @@
+//|     from binascii import hexlify
+//|
+//|     key = b'Sixteen byte key'
+//|     cipher = new aes(key, aes.mode.MODE_ECB)
```
```suggestion
//|     cipher = aes.AES(key, aes.mode.MODE_ECB)
```

tannewt commented on shared-bindings/aes/aes.c:
```diff
@@ -0,0 +1,355 @@
+
+  // Set defaults. These will be overridden in mp_arg_parse_all() if an
+  // argument is provided.
+  args[ARG_mode].u_int = 0;
```
Defaults should be placed in allowed_args above if they aren't just 0. Here is an example: https://github.com/adafruit/circuitpython/blob/master/shared-bindings/busio/I2C.c#L74

tannewt commented on shared-bindings/aes/aes.c:
```diff
@@ -0,0 +1,355 @@
+//|
+//| An object that represents an AES stream, including the current state.
+//|
+//| .. class:: AES(key[, mode, iv])
```
```suggestion
//| .. class:: AES(key, mode=0, iv=None, segment_size=8)
```

tannewt commented on shared-bindings/aes/aes.c:
```diff
@@ -0,0 +1,355 @@
+        translate("Source and destination buffers must be the same length"));
+  }
+
+  memcpy(destbufinfo.buf, srcbufinfo.buf, srcbufinfo.len);
```
Why do this? Isn't it simpler to have a dual buffer implementation under the hood and pass it twice for in-place?

xobs replied:
The actual library call takes the form of `AES_CBC_decrypt_buffer(struct AES_ctx* ctx, uint8_t* buf, uint32_t length);` or even `AES_ECB_decrypt(const struct AES_ctx* ctx, uint8_t* buf);`.  Internally it's a single buffer, so we need to do the copy.

tannewt commented on shared-bindings/aes/aes.c:
```diff
@@ -0,0 +1,355 @@
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(aes_encrypt_into_obj, aes_encrypt_into);
+
+//|   .. method:: encrypt(buf)
+//|
+//|      Encrypt the buffer and return a copy of the encrypted data.
+//|      The buffer must be a multiple of 16 bytes.
+//|
```
I wouldn't even provide this. Folks should manage the buffers themselves.

xobs replied:
This comes from when we were going for `pycrypto` compatibility.  Since that's no longer a goal, I'll remove it.

tannewt commented on shared-bindings/aes/aes.c:
```diff
@@ -0,0 +1,355 @@
+MP_DEFINE_CONST_FUN_OBJ_VAR(aes_rekey_obj, 2, aes_rekey);
+
+
+STATIC mp_obj_t aes_set_mode(mp_obj_t aes_obj, mp_obj_t mode_obj) {
```
Please add mode as a property instead of having a `set_mode` function. Here is an example: https://github.com/adafruit/circuitpython/blob/master/shared-bindings/busio/SPI.c#L372-L389

tannewt commented on shared-module/aes/__init__.c:
```diff
@@ -0,0 +1,82 @@
+                            size_t length) {
+  switch (self->mode) {
+    case AES_MODE_ECB:
+      if (length != 16) {
+        mp_raise_msg(&mp_type_ValueError, translate("ECB only operates on 16 bytes at a time"));
+      }
```
Is this an implementation detail? If not, then shared-bindings should check it.

tannewt commented on shared-module/aes/__init__.c:
```diff
@@ -0,0 +1,82 @@
+      AES_CTR_xcrypt_buffer(&self->ctx, buffer, length);
+      break;
+    default:
+      mp_raise_msg(&mp_type_ValueError, translate("Unknown encryption mode"));
```
You shouldn't need this here. The constructor and mode set should have verified the value already.

xobs commented:
Changes:

* Rebased onto current master
* Removed non-`_into` routines (encrypt, decrypt, encrypt_in_place, and decrypt_in_place)
* Renamed to `aesio`
* Fixed documentation as requested
* Removed checks from the `shared-modules`
* Made `mode` a property
* Moved checks into their own function
* Enable variable-length CTS operations (previously was limited to `% 16` bytes)
